File: 0127-word-ladder/0127-word-ladder.py
import collections

class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        if endWord not in wordList:
            return 0
        
        wordList.append(beginWord)
        graph = collections.defaultdict(list)
        
        def make_graph():
            for wrd in wordList:
                for i in range(len(wrd)):
                    x = wrd[:i] + "*" + wrd[i+1:]
                    graph[x].append(wrd)

        make_graph()

        # BFS
        def bfs():
            queue = collections.deque([(beginWord, 1)])
            visited = set()
            visited.add(beginWord)

            while queue:
                current_word, level = queue.popleft()
                
                for i in range(len(current_word)):
                    intermediate_word = current_word[:i] + "*" + current_word[i+1:]
                    for neighbor in graph[intermediate_word]:
                        if neighbor == endWord:
                            return level + 1
                        if neighbor not in visited:
                            visited.add(neighbor)
                            queue.append((neighbor, level + 1))
            return 0

        return bfs()

File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    vector<string> createDiff(string a){
        vector<string> res;
        string alph = "abcdefghijklmnopqrstuvwxyz";

        char rep;
        for (auto& ch: a){
            rep = ch;

            for (auto el: alph){
                ch = el;
                res.push_back(a);
                ch = rep;
            }
        }

        return res;
    }

    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        set<string> s(wordList.begin(), wordList.end());
        queue<pair<string, int>> q;
        q.push({beginWord, 1});

        while (!q.empty()){
            auto [str, step] = q.front();
            q.pop();

            if (str == endWord) return step;

            for (auto el: createDiff(str)){
                if (s.find(el) != s.end()){
                    s.erase(el);
                    q.push({el, step + 1});
                }
            }
        }

        return 0;
    }
};